1-Draft/RFC0000-Chain-Operators.md

```diff
@@ -0,0 +1,565 @@
+---
+RFC:
+Author: Robert Holt
+Status: Draft
+Area: Pipeline Chain Operators
+Comments Due: 2019-07-12 0000
+Plan to implement: Yes
+---
+
+# Pipeline Chain Operators
+
+POSIX shells have what may be referred to as *AND-OR lists*,
+what the [Open Group's Shell Command Language specification](https://pubs.opengroup.org/onlinepubs/007904875/utilities/xcu_chap02.html#tag_02_09_03)
+describes as:
+
+> a sequence of one or more pipelines separated by the operators `&&` and `||`
+
+Semantically, these operators look at the exit code of the left-hand side pipeline
+and conditionally execute the right-hand side pipeline.
+`&&` executes the right-hand side if the left-hand side "succeeded",
+and `||` executes the right-hand side if the left-hand side did not "succeed".
+
+Some examples in bash:
+
+```bash
+echo 'Success' && echo 'Second success' # Executes both left-hand and right-hand command
+echo 'Success' || echo 'Second success' # Executes only the left-hand command
+false && echo 'Second success' # Executes only the left-hand command (false always "fails")
+false || echo 'Second success' # Executes both left-hand and right-hand commands
+false && echo foo || echo bar # Writes only "bar" to stdout (operators are left-associative)
+true || echo foo && echo bar # Writes only "bar" to stdout
+```
+
+Also from the Shell Command Language specification:
+
+> A ';' or &lt;newline&gt; terminator shall cause the preceding AND-OR list to be executed sequentially;
+> an '&' shall cause asynchronous execution of the preceding AND-OR list.
+
+Meaning that the entire AND-OR list is backgrounded:
+
+```bash
+true && echo 'Success' & # The list `true && echo 'Success'` is executed in the background
+```
+
+This RFC proposes adding AND-OR lists to PowerShell, using the same `&&` and `||`.
+Because "AND-OR list" is neither intuitive nor common as a terminology,
+instead the term **pipeline chains** is proposed.
+
+## Motivation
+
+> As a PowerShell user, I can chain commands with `&&` and `||`
+> so that I have an ergonomic syntax to conditionally invoke side-effectful commands.
+

```

Comment:
Ha!

Well they're also present in cmd, but really I'm a PowerShell user and I just want these in PowerShell so I can do stuff like `Import-Module ./build.psm1 && Start-PSBuild && Start-PSPester` and `sudo apt update && sudo apt upgrade`.

Reply:
But my thinking on this feature is not to make PowerShell more bash-like. The operators are nice, but they must be PowerShell's own. Taking an operator doesn't mean we should take all the semantics.

Reply:
@KirkMunro Is it so wrong to make it easier for bash users to adopt PowerShell? It seems that you are saying that with a derisive attitude rather than a positive one. I think it's great to make it as easy as possible for people to adapt to powershell, and they days of Microsoft's exclusive attitude are gone, they "love linux".

In addition, it's not just a cosmetic issue, there are very real and valuable benefits to the `&&` and `||` operators that are a pain to work around, somewhat less intuitive, and harder/more to remember.

For example, I'm currently building a command line tool, and I want it to compile, and if the compilation works, then run the command, like so:

```
go install && mycmd run
```

Pretty simple right? Easy/quick to type, and it's very clear and easy to read.

With the current implementation of powershell though, I would have to do something like this:

```
$(command -arg1 -arg2 | Out-Host;$?) -and $(command2 -arg1 | Out-Host;$?)
```

That's a lot to type, when you just want to run a quick command, and there's really no reason for it to be that way. I'm not saying you couldn't develop a workaround, but I just don't see why anyone on God's green earth would be against making this simpler.

Reply:
Not to mention, you have to run it in a sub shell and don't get to see the output of your commands, even with the overly verbose version

Reply:
@zwhitchcox I wasn't being derisive, and I fully embrace making PowerShell more familiar for bash users, C# programmers, etc.

The comment I made was meant to suggest that we could just literally state the motivation behind this RFC. Nothing more was behind it than that.

Reply:
@KirkMunro Sorry, I think I misinterpreted your meaning, after I read the rest of this forum, it was clear you weren't trying to exclude bash users